File: Python/src/day_1.py
from src import utils
import numpy as np

class Day1:
    def __init__(self):
        self.file_contents = utils.read_file ("./files/day_1.txt")
        self.left_numbers = []
        self.right_numbers = []

        self.left_numbers, self.right_numbers = self.split_nums(self.file_contents)

    @staticmethod
    def split_nums(content: str):
        line_pair = content.splitlines()

        left_numbers = []
        right_numbers = []

        for n in line_pair:
            left, right = n.split("   ")
            left_numbers.append(int(left))
            right_numbers.append(int(right))
        
        left_numbers.sort()
        right_numbers.sort()

        assert len(left_numbers) == len(right_numbers)
        return left_numbers, right_numbers

    def run_part_1(self):
        diff = []

        for i in range(len(self.left_numbers)):
            diff.append (abs(self.left_numbers[i] - self.right_numbers[i]))

        result = np.sum (diff)
        print (result) 
    
    def run_part_2(self):
        similarity = []
        
        for l_number in self.left_numbers:
            matching_numbers = [x for x in self.right_numbers if x == l_number]
            similarity.append (l_number * len(matching_numbers))
            
        result = np.sum (similarity)
        print (result)

File: cpp/Day1.cpp
#include "Day1.h"

#include <cstdio>
#include <fstream>
#include <iostream>
#include <numeric>

Day1::Day1()
{
    std::cout << "Day 1" << std::endl;

    std::filesystem::path contentPath = "/Users/reubenduckering/Documents/Personal Repo/Advent-Of-Code-2024/files/day_1.txt";

    std::ifstream contentFile (contentPath);

    std::string myText;

    while (getline (contentFile, myText))
    {
        std::string leftNumber = myText.substr(0, myText.find("   "));
        std::string rightNumber = myText.substr(myText.find("   ") + 3, myText.length());

        leftNumbers.push_back (std::stoi (leftNumber));
        rightNumbers.push_back (std::stoi (rightNumber));
    }

    contentFile.close();

    std::sort (leftNumbers.begin(), leftNumbers.end());
    std::sort (rightNumbers.begin(), rightNumbers.end());
}

void Day1::runPart1()
{
    std::vector<int> diff;
    diff.resize (leftNumbers.size());

    for (int i = 0; i < leftNumbers.size(); ++i)
        diff.push_back (abs (leftNumbers[i] - rightNumbers[i]));

    int result = std::accumulate (diff.begin(), diff.end(), 0);

   printf ("Result: %d\n", result);
}

void Day1::runPart2()
{
    std::vector<int> similarity;
    similarity.resize (leftNumbers.size());

    for (const auto& lNum: leftNumbers)
    {
        int cnt = 0;

        for (const auto& rNum: rightNumbers)
        {
            if (lNum == rNum)
                ++cnt;

            // It's sorted so return if the number is largest as they'll be no more.
            if (rNum > lNum)
                break;
        }

        similarity.push_back (lNum * cnt);
    }

    int result = std::accumulate (similarity.begin(), similarity.end(), 0);

    printf ("Result: %d\n", result);
}